server/queue/queue-pulsar/src/main/scala/org/apache/james/queue/pulsar/PulsarMailQueue.scala

```diff
@@ -0,0 +1,587 @@
+        .filter(_.messageId.underlying <= lastMessageId.underlying) //stop copying message at lastMessageId
+        .runFold(0L) { (counter, message) =>
+          putMessageInOutTopic(message)
+          counter + 1
+        }
+      Await.result(copy, Duration.Inf)
+    }
+  }
+
+  private def read(config: ReaderConfig)(implicit executionContext: ExecutionContext): Source[ConsumerMessage[String], NotUsed] =
+    Source.unfoldResourceAsync[ConsumerMessage[String], Reader[String]](
+      create = () => {
+        Future.successful(
+          client.reader(
+            config = config
+          )
+        )
+      },
+      read = reader => {
+        if (reader.hasMessageAvailable) reader.nextAsync.map(Some(_))
+        else Future.successful(None)
+      },
+      close = reader => reader.closeAsync.map(_ => Done))
+
+
+  /**
+   * @inheritdoc
+   */
+  override def clear(): Long = {
+    val count = getSize()
+    admin.topics().delete(outTopic.name, true)
+    count
+  }
+
+  private def lastMessage(topic: Topic): Option[ConsumerMessage[String]] = {
+    val reader = client.reader(
+      config = ReaderConfig(topic,
+        startMessage = Message(MessageId.latest),
+        startMessageIdInclusive = true)
+    )
+
+    if (reader.hasMessageAvailable)
+      reader.next(10.seconds)
+    else
+      None
+  }
+
+  /**
+   * Remove all mails from the queue that match
+   *
+   * The intent of remove is to allow operators to clear some emails from the mailqueue
+   * in an emergency situation such as a DOS attempt or a configuration error which creates a mail loop
+   * or a bounce loop.
+   *
+   * The chosen implementation is to insert a filter which will very quickly dequeue matching emails while continuing
+   * to process normal traffic. This means that the count() method will not immediately account for the removed emails
+   * but will eventually converge once all the matching emails in the mailqueue have been processed.
+   *
+   * We will use actor interop patterns from akka stream to interoperate with a filtering actor
+   * https://doc.akka.io/docs/akka/current/stream/actor-interop.html
+   *
+   * @return 0 because this is an async and distributed process running on several nodes on the cluster
+   */
+  override def remove(`type`: ManageableMailQueue.Type, value: String): Long = {
+    import Filter._
+
+    val maybeLastSequenceId = lastMessage(outTopic).map(_.sequenceId)
+
+    maybeLastSequenceId.foreach { lastSequenceId =>
+      import ManageableMailQueue.Type
+      val filter: Filter = `type` match {
+        case Type.Sender =>
+          Filter.BySender(value, lastSequenceId)
+        case Type.Recipient =>
+          Filter.ByRecipient(value, lastSequenceId)
+        case Type.Name =>
+          Filter.ByName(value, lastSequenceId)
+      }
+      filterProducer.send(Json.stringify(Json.toJson(filter)))
+    }
+    0
```

Comment:
We are always returning 0?

Reply:
Yes, it is. Documented here: https://github.com/apache/james-project/pull/808/files/e53fd053665ba0ebfd71258f94dd46dee5a568a6#diff-fff51bf28740b45820cda4a05033a6642999532534a71e7e33d43fae17f94c35R475

The rational is: deletion will happen by filtering the mailqueue on dequeue (by design), so it's impossible to provide a decent count here.